lib/classes/EnvLoader.js

```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+const dotenv = require('dotenv');
+const fileExists = require('../utils/fs/fileExists');
+const path = require('path');
+const _ = require('lodash');
+
+class EnvLoader {
```

Comment:
I was also not sure but I decided to follow similar approach as with `YamlParser` class. I currently changed implementation from class-based to function-based, however, it still depends on `serverless` instance. I was thinking on how to separate it, but it actually requires a few dependencies such as access to `serverlessConfigFile`, `options` and `logDeprecation`. I'm on the fence here and I see a few options:
- keep the dependency on `serverless`
- keep the dependency on specific pieces - instead of passing `serverless` instance, just pass the needed pieces
- resolve everything in `Serverless` and only pass `stage`, `useDotenv` to `loadEnv` function

Which one do you think makes the most sense here? I personally like the option #3, but that moves some of the responsibilities of `loadEnv` directly to `Serverless` class, which is not ideal

Reply:
> I was also not sure but I decided to follow similar approach as with YamlParser class

Totally understood, we have a lot of legacy conventions. Still it's good to abort some of them, especially when they resemble noisy not well justified practices.

> it still depends on `serverless` instance

That's true. I think the cleanest approach for now, would be to follow with something as:

_In`Serverless.js` (inline in `init` or in new method):_

```javascript
if (serviceContext && serviceConfig.useDotEnv) {
  const stage = // resolve stage
  loadDotEnv(stage);
} else {
  isDotFileEnvExisting = // fs exists checks
  if (isDotFileEnvExisting) showDeprecation();
}
```

And in _lib/resolveDotEnv.js_  just load eventually existing config for given stage

Note, that if we do it that way, then we would be able to use `lib/resolveDotEnv.js` in as is form, in refactor we plan to do here: https://github.com/serverless/serverless/issues/8364  (it is mostly about separating CLI and programmatic logic, so Framework core is purely programmatic (there will be then no dotenv resolution in its core), it'll also allow to integrate Components better)

Reply:
So yes, option 3 as you pointed, seems best :)

Reply:
But without passing `useDotenv`  (let's call the util when we know we want to resolve env from `.env`)

Reply:
Thanks for a comprehensive explanation, I wasn't sure if moving some of the env-related logic makes sense in this case, but thanks to your explanation I see the reasoning and I also think that this is the most appropriate solution moving forward :bow: